Implement binding and lowering for InterpolatedStringBuilder

333fred opened the pull request:
This implements initial binding and lowering for the InterpolatedStringBuilder, used when an interpolated string is converted directly to a `string`, and not to some other builder type. I've written it in such a way as to make future refactorings to support general builder types simple, but some of the steps (particularly lowering) are more simplistic for now. I did _not_ implement nullable analysis or definite assignment in this PR, prototype comments have been left to make sure that those are handled later. The spec for this feature is located at https://github.com/dotnet/csharplang/blob/main/proposals/improved-interpolated-strings.md, I will have a PR out later today or tomorrow to reflect the changes from LDM earlier today.

333fred commented:
@chsienki @dotnet/roslyn-compiler for review.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -376,14 +376,7 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, BindingDi
-                            unconvertedInterpolatedString.ConstantValue,
-                            unconvertedInterpolatedString.Type,
-                            unconvertedInterpolatedString.HasErrors);
+                        result = BindUnconvertedInterpolatedStringToString(unconvertedInterpolatedString, diagnostics);
```
nit: naming. We're not returning a string. Consider just `BindUnconvertedInterpolatedString`

333fred replied:
This naming is intentional. We're binding this interpolated string to `System.String`, not to some builder type. We'll eventually have a `BindUnconvertedInterpolatedStringToBuilderType` as well.

---
In reply to: [608901424](https://github.com/dotnet/roslyn/pull/52422#discussion_r608901424) [](ancestors = 608901424)

chsienki commented on src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs:
```diff
@@ -529,7 +529,25 @@ public void CloseMethod(BoundStatement body)
+
+        public LocalSymbol InterpolatedStringBuilderLocal(
+            TypeSymbol type,
+            uint valEscapeScope,
```
Consider just adding valEscapeScope as a default parameter to `SynthesizedLocal`. If the escape is _only_ ever going to be applicable to the the string builder, perhaps we should represent it with a different symbol type.

333fred replied:
I waffled back and forth on whether to define a separate type or just use `SynthesizedLocal`. I can break out it into a separate type.

---
In reply to: [608906130](https://github.com/dotnet/roslyn/pull/52422#discussion_r608906130) [](ancestors = 608906130)

chsienki commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedLocal.cs:
```diff
@@ -48,13 +50,16 @@ internal sealed class SynthesizedLocal : LocalSymbol
             Debug.Assert(!type.IsVoidType());
             Debug.Assert(!kind.IsLongLived() || syntaxOpt != null);
             Debug.Assert(refKind != RefKind.Out);
+            Debug.Assert((valEscapeScope == null && kind != SynthesizedLocalKind.InterpolatedStringBuilder) ||
```
Hmm, so it looks like we are only allowing escape for string builders. This feels... clunky? to me? That we have a single type that reprsents two different things. Although I'm not sure if it's worth splitting it up or not...

chsienki commented on src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs:
```diff
@@ -194,7 +194,12 @@ private void CaptureVariable(Symbol variable, SyntaxNode syntax)
-                (_lazyDisallowedCaptures ??= new MultiDictionary<Symbol, SyntaxNode>()).Add(variable, syntax);
+                // Interpolated string builder variables are limited in lifetime, and will not be used in a context that
+                // actually needs to be saved to the closure.
+                if (variable is not SynthesizedLocal { SynthesizedKind: SynthesizedLocalKind.InterpolatedStringBuilder })
```
Another case where we're treating the synthesized locals differently depending on the underlying kind :/

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,278 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                var needToCheckConversionToObject = !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
+                var (objectType, conversionDiagnostics) = needToCheckConversionToObject
+                    ? (GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax), BindingDiagnosticBag.GetInstance())
+                    : (null, null);
```
nit: this is somewhat hard to follow (in my head at least). Consider rewriting this as 

```chsarp 
SpecialType objectType = null;
if (!Compilation.IsFeatureEnabled() ...)
{
   objectType = GetSpecialType...
}
```

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,278 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    {
+                        numFormatHoles++;
+
+                        if (needToCheckConversionToObject)
```
Consider extracting this logic to another local function

333fred replied:
I'm going to hold off on this for now, as I think this code will need to be updated to get a better error experience for things that cannot be in interpolated string holes (such as `ReadOnlySpan<char>`

---
In reply to: [609046151](https://github.com/dotnet/roslyn/pull/52422#discussion_r609046151) [](ancestors = 609046151)

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,278 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                if (disposeMethod != null)
+                {
+                    var argsToParams = default(ImmutableArray<int>);
+                    var expanded = disposeMethod.HasParamsParameter();
```
If the disposeMethod is well known, how would it ever have params parameters? Generally this is only needed for pattern-based dispose lookup, which I don't believe we opted to support?

333fred replied:
The prototype comment above kinda covers this. We likely want to support pattern-based dispose in the general case.

---
In reply to: [609073067](https://github.com/dotnet/roslyn/pull/52422#discussion_r609073067) [](ancestors = 609073067)

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,278 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    var argsToParams = default(ImmutableArray<int>);
+                    var expanded = disposeMethod.HasParamsParameter();
+
+                    BindDefaultArguments(
```
Same again, it feels weird to be attempting to bind default arguments for a method we wouldn't expect to have them?

333fred replied:
I can probably just remove this part since we match by well-known type, and we'll want to use the existing machinery for the general case if we decide to support it.

---
In reply to: [609079627](https://github.com/dotnet/roslyn/pull/52422#discussion_r609079627) [](ancestors = 609079627)

chsienki commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,278 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    ReportDiagnosticsIfUnmanagedCallersOnly(diagnostics, createMethod, unconvertedInterpolatedString.Syntax.Location, isDelegateConversion: false);
+                }
+
+                // PROTOTYPE(interp-strings): Do we need to support versions of InterpolatedStringBuilder that do not have Dispose, or implement it
```
Aside: If we do go the route of allowing optional/pattern based dispose lookup, we can probably re-use the already implemented logic.

333fred replied:
I had actually made the changes to the code to support it, but rolled it back until we figure out the broader dispose story.

---
In reply to: [609084092](https://github.com/dotnet/roslyn/pull/52422#discussion_r609084092) [](ancestors = 609084092)

chsienki commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+
+            var typeArguments = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+            var implicitBuilderReceiver = new BoundImplicitReceiver(source.Syntax, builderType);
+            var analyzedArguments = AnalyzedArguments.GetInstance();
```
Is this Free'd anywhere?

chsienki commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+
+                _binder.CoerceArguments(overloadResolutionResult.ValidResult, analyzedArguments.Arguments, diagnostics);
+
+                bool expanded = overloadResolutionResult.ValidResult.Result.Kind == MemberResolutionKind.ApplicableInExpandedForm;
```
So the spec above says 'overload resolution contains a single result'... which matches here. I guess I'm suprised that we're allowing the candidate methods to include params and default args, etc, given that we're fairly prescriptive about the rest of the shape. Is there an explicit use case for this?

333fred replied:
> I guess I'm suprised that we're allowing the candidate methods to include params and default args, etc, given that we're fairly prescriptive about the rest of the shape.

We're not that prescriptive. I've taken a couple of shortcuts because of the specific type involved with interpolated strings as strings, but in general we'll have this type of support for `Create` and `Dispose` (if we support dispose at all).

---
In reply to: [609090339](https://github.com/dotnet/roslyn/pull/52422#discussion_r609090339) [](ancestors = 609090339)

333fred replied:
As to use cases: any of the caller info attributes, for example.

---
In reply to: [609114486](https://github.com/dotnet/roslyn/pull/52422#discussion_r609114486) [](ancestors = 609114486,609090339)

chsienki commented on src/Compilers/VisualBasic/Test/Syntax/IncrementalParser/IncrementalParser.vb:
```diff
@@ -2602,7 +2602,7 @@ End Class
             If (c = vbLf(0)) AndAlso
                 ((i = str.Length - 1) OrElse (str(i + 1) <> vbCr(0))) Then
-                builder.AddRange(vbCrLf)
+                builder.AddRange(DirectCast(vbCrLf, IEnumerable(Of Char)))
```
Why this change?

333fred replied:
Oh, this one can be reverted, thanks. I had added an ambiguous overload to `ArrayBuilder` with `ReadOnlySpan`, but had to revert it.

---
In reply to: [609090644](https://github.com/dotnet/roslyn/pull/52422#discussion_r609090644) [](ancestors = 609090644)

chsienki reviewed (dismissed):
:clock1:

333fred commented:
@chsienki addressed feedback.

chsienki reviewed (dismissed):
:shipit:

333fred commented:
@dotnet/roslyn-compiler for a second review.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -2014,6 +2014,11 @@
+      <Field Name="InterpolationData" Type="InterpolatedStringBuilderData?"/>
+  </Node>
+    
+  <Node Name="BoundInterpolatedStringElementPlaceholder" Base="BoundExpression">
```
>BoundExpression [](start = 63, length = 15)

```BoundValuePlaceholderBase```?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/InterpolatedStringBuilderData.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed record InterpolatedStringBuilderData(
```
>InterpolatedStringBuilderData [](start = 27, length = 29)

What record features do we actually take advantage of here? Given how much metadata is going to be produced for this type, we better use all of it. Otherwise, I think we should find another alternative. We also should review other "records" recently added to the compiler.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/InterpolatedStringBuilderData.cs:
```diff
@@ -0,0 +1,16 @@
+        MethodArgumentInfo? Construction,
+        ImmutableArray<MethodArgumentInfo> BuilderFormatCalls,
+        MethodArgumentInfo? DisposeInfo,
+        uint ScopeOfContainingExpression);
```
>ScopeOfContainingExpression [](start = 13, length = 27)

What exectly is this?

333fred replied:
I'll add a doc comment. It's the scope current expression during binding, for the purposes of setting the ValEscapeScope of the interpolated string builder.

---
In reply to: [610639085](https://github.com/dotnet/roslyn/pull/52422#discussion_r610639085) [](ancestors = 610639085)

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:
```diff
@@ -66,5 +66,31 @@ public static bool IsConstructorInitializer(this BoundCall call)
+        {
+            public bool ContainsAwait = false;
+
+            public override BoundNode? VisitAwaitExpression(BoundAwaitExpression node)
```
>VisitAwaitExpression [](start = 39, length = 20)

Consider also overriding non-strongly type Visit to short-circuit if we already found an await.

AlekseyTs commented on src/Compilers/Core/Portable/SynthesizedLocalKind.cs:
```diff
@@ -232,6 +232,12 @@ internal enum SynthesizedLocalKind
         DelegateRelaxationReceiver = 0x101,
+
+        /// <summary>
+        /// The builder type of an interpolated string, when using the interpolated string builder pattern.
```
>The builder type of an interpolated string, when using the interpolated string builder pattern. [](start = 12, length = 95)

The relationship between the builder type and the local is not clear. Is this statement talking about the type of the local?

333fred replied:
Rewording this for clarity.

---
In reply to: [610649620](https://github.com/dotnet/roslyn/pull/52422#discussion_r610649620) [](ancestors = 610649620)

AlekseyTs commented on src/Compilers/Core/Portable/SynthesizedLocalKind.cs:
```diff
@@ -232,6 +232,12 @@ internal enum SynthesizedLocalKind
+        /// The builder type of an interpolated string, when using the interpolated string builder pattern.
+        /// This synthesized local is allowed to have a val escape
+        /// </summary>
+        InterpolatedStringBuilder = 0x102,
```
>InterpolatedStringBuilder [](start = 8, length = 25)

I performed a search for "InterpolatedStringBuilder.SynthesizedLocalKind" in files in this PR. CodeFlow shows only one match. It appears no one is specifically checking for this value. It is quite possible that the textual search doesn't find all references for this member, but, if it does, it doesn't look like there is a real reason behind adding this member (at least in this PR).

333fred replied:
After the refactor, I am using it.

---
In reply to: [610658103](https://github.com/dotnet/roslyn/pull/52422#discussion_r610658103) [](ancestors = 610658103)

333fred replied:
After the full refactor and addressing your comments, I'm not using this value directly, so I _could_ turn it back into a local rewriting temp. However, I'd prefer to keep it as a distinct value, for my ease of debugging.

---
In reply to: [612030082](https://github.com/dotnet/roslyn/pull/52422#discussion_r612030082) [](ancestors = 612030082,610658103)

AlekseyTs commented on src/Compilers/Core/Portable/SynthesizedLocalKind.cs:
```diff
@@ -232,6 +232,12 @@ internal enum SynthesizedLocalKind
+        /// The builder type of an interpolated string, when using the interpolated string builder pattern.
+        /// This synthesized local is allowed to have a val escape
+        /// </summary>
+        InterpolatedStringBuilder = 0x102,
```
>InterpolatedStringBuilder [](start = 8, length = 25)

Is this just a specialized flavor of ```LoweringTemp```? I.e. it is "allowed to have a val escape" (BTW, it would be good to provde a brief explanation what does that mean exactly). Perhaps we should use a more general name rather than tying it to specific feature and specific use case by name.

333fred replied:
Yes. I originally had just modified `LoweringTemp`, but got feedback that it might be better as a separate class.

---
In reply to: [610661871](https://github.com/dotnet/roslyn/pull/52422#discussion_r610661871) [](ancestors = 610661871)

333fred replied:
> Perhaps we should use a more general name rather than tying it to specific feature and specific use case by name.

I like this suggestion, I'll do that.

---
In reply to: [612025077](https://github.com/dotnet/roslyn/pull/52422#discussion_r612025077) [](ancestors = 612025077,610661871)

333fred replied:
I've created a more general `SynthesizedLocalWithValEscape`, but left this kind as is for ease of debugging.

---
In reply to: [612025148](https://github.com/dotnet/roslyn/pull/52422#discussion_r612025148) [](ancestors = 612025148,612025077,610661871)

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>Using a function pointer type in a 'typeof' in an attribute is not supported.</value>
   </data>
+  <data name="IDS_FeatureImprovedInterpolatedStrings" xml:space="preserve">
+    <value>improved interpolated strings</value>
```
>improved [](start = 11, length = 8)

How are we going to call the next improvement to interpolated strings? Consider using more specific feature name.

333fred replied:
I'll rename to `interpolated string builders`

---
In reply to: [610676632](https://github.com/dotnet/roslyn/pull/52422#discussion_r610676632) [](ancestors = 610676632)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/InterpolatedStringBuilderLocalSymbol.cs:
```diff
@@ -0,0 +1,116 @@
+            uint valEscapeScope
+#if DEBUG
+            ,
+            int createdAtLineNumber,
```
>createdAtLineNumber [](start = 16, length = 19)

is this supposed to be a```[CallerLineNumber]``` parameter?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/InterpolatedStringBuilderLocalSymbol.cs:
```diff
@@ -0,0 +1,116 @@
+    /// A synthesized local variable.
+    /// </summary>
+    [DebuggerDisplay("{GetDebuggerDisplay(),nq}")]
+    internal sealed class InterpolatedStringBuilderLocalSymbol : LocalSymbol
```
>InterpolatedStringBuilderLocalSymbol [](start = 26, length = 36)

Is this mostly a clone of ```SynthesizedLocal``` type? In that case I would prefer us to either inherit from ```SynthesizedLocal```, or inherit from a common base extracted from ```SynthesizedLocal``` without changing style for properties to simplify the review process.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs:
```diff
@@ -194,7 +194,12 @@ private void CaptureVariable(Symbol variable, SyntaxNode syntax)
-                (_lazyDisallowedCaptures ??= new MultiDictionary<Symbol, SyntaxNode>()).Add(variable, syntax);
+                // Interpolated string builder variables are limited in lifetime, and will not be used in a context that
+                // actually needs to be saved to the closure.
+                if (variable is not InterpolatedStringBuilderLocalSymbol)
```
Why can't we just let the machinery to figure this out.

333fred replied:
I'm honestly not sure what you mean by this. What machinery are you referring to?

---
In reply to: [610693222](https://github.com/dotnet/roslyn/pull/52422#discussion_r610693222) [](ancestors = 610693222)

AlekseyTs replied:
> I'm honestly not sure what you mean by this. What machinery are you referring to?

Given that "and will not be used in a context that actually needs to be saved to the closure", why do we need this special case at all? 

---
In reply to: [612033020](https://github.com/dotnet/roslyn/pull/52422#discussion_r612033020) [](ancestors = 612033020,610693222)

333fred replied:
I think this was due to the disposal mechanisms. As that's no longer present, the machinery no longer needs to be adjusted.

---
In reply to: [612764219](https://github.com/dotnet/roslyn/pull/52422#discussion_r612764219) [](ancestors = 612764219,612033020,610693222)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:
```diff
@@ -74,6 +73,10 @@ internal ConversionsBase WithNullability(bool includeNullability)
         protected abstract Conversion GetInterpolatedStringConversion(BoundUnconvertedInterpolatedString source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo);
 
+#nullable  enable
+        public abstract bool IsApplicableInterpolatedStringBuilderType(BoundUnconvertedInterpolatedString source, TypeSymbol builderType, BindingDiagnosticBag diagnostics, [NotNullWhen(true)] out ImmutableArray<MethodArgumentInfo> builderArguments);
```
>[NotNullWhen(true)] [](start = 172, length = 19)

The type of the parameter is not nullable.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+        {
+            // SPEC:
+            // A type is said to be an _applicable_interpolated_string_builder_type_ if, given an _interpolated_string_literal_ `S`, the following is true:
+            //  * Overload resolution with an identifier of `TryFormatBaseString` and a parameter type of `string` succeeds, and contains a single instance method that returns a `bool` or `void`.
```
>Overload resolution with an identifier of `TryFormatBaseString` and a parameter type of `string` succeeds, and contains a single instance method that returns a `bool` or `void`. [](start = 18, length = 177)

Consider shortening the lines in the comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+            //    contains a single instance method that returns a `bool` or `void`.
+            // Additionally, all resolved method calls must return the same type. `TryFormat` calls that mix `bool` and `void` are not permitted.
+
+            // If the type is applicable, we'll add these to the given info at the end.
```
>If the type is applicable, we'll add these to the given info at the end. [](start = 15, length = 72)

It doesn't look like this is accurate. Also, I think it better to allways add use site info, even in case of a failure because diagnostics might provide a hint why things do not work.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
         }
 
+#nullable  enable
+        public override bool IsApplicableInterpolatedStringBuilderType(BoundUnconvertedInterpolatedString source, TypeSymbol builderType, BindingDiagnosticBag diagnostics, out ImmutableArray<MethodArgumentInfo> builderArguments)
```
>builderArguments [](start = 211, length = 16)

Are these really arguments? I think we need to come up with a name that more accurately reflects the nature of data.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var overloadResolutionResult = OverloadResolutionResult<MethodSymbol>.GetInstance();
+            var stringType = _binder.GetSpecialType(SpecialType.System_String, diagnostics, source.Syntax);
+            analyzedArguments.Arguments.Add(new BoundLiteral(CSharpSyntaxTree.Dummy.GetRoot(), constantValueOpt: null, stringType));
```
>new BoundLiteral(CSharpSyntaxTree.Dummy.GetRoot(), constantValueOpt: null, stringType) [](start = 44, length = 86)

It doesn't feel like a string null literal is an appropriate argument to use. The spec says a string value, not a null literal.

AlekseyTs replied:
Don't we already have the actual value we are planning to pass to the method?

---
In reply to: [610718396](https://github.com/dotnet/roslyn/pull/52422#discussion_r610718396) [](ancestors = 610718396)

333fred replied:
> Don't we already have the actual value we are planning to pass to the method?

If there is no content in the string, then no.

---
In reply to: [610719208](https://github.com/dotnet/roslyn/pull/52422#discussion_r610719208) [](ancestors = 610719208,610718396)

AlekseyTs replied:
>> Don't we already have the actual value we are planning to pass to the method?

> If there is no content in the string, then no.

Then why do we care about this method at all? Even if we do, I do not believe using a null literal is the right thing to do.

---
In reply to: [612033734](https://github.com/dotnet/roslyn/pull/52422#discussion_r612033734) [](ancestors = 612033734,610719208,610718396)

333fred replied:
> Then why do we care about this method at all?

The spec mandates that this method exists at least. Generally, I'm trying to avoid `$""` being convertible to something that really wasn't intended to be a builder and happens to have a method named `Create`, takes 2 int parameters, and returns the current type. That is likely to be a somewhat common thing. We could also word the spec a bit differently to say that member lookup with `AppendLiteral` and `AppendFormat` must have method members?

---
In reply to: [612766918](https://github.com/dotnet/roslyn/pull/52422#discussion_r612766918) [](ancestors = 612766918,612033734,610719208,610718396)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var overloadResolutionResult = OverloadResolutionResult<MethodSymbol>.GetInstance();
+            var stringType = _binder.GetSpecialType(SpecialType.System_String, diagnostics, source.Syntax);
+            analyzedArguments.Arguments.Add(new BoundLiteral(CSharpSyntaxTree.Dummy.GetRoot(), constantValueOpt: null, stringType));
```
>CSharpSyntaxTree.Dummy.GetRoot() [](start = 61, length = 32)

```source.Syntax```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                // PROTOTYPE(interp-strings): We'll want to have a specific error for when we're converting to a non-string type
+                free();
+                builderArguments = default;
+                return false;
```
>return false; [](start = 16, length = 13)

Dropping useSiteInfo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                default:
+                    free();
+                    builderArguments = default;
+                    return false;
```
>return false; [](start = 20, length = 13)

Dropping useSiteInfo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                    builderFormatCalls.Free();
+                    interpolationHoleCandidates?.Free();
+                    builderArguments = default;
+                    return false;
```
>return false; [](start = 20, length = 13)

Dropping useSiteInfo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                }
+                else
+                {
+                    candidateMethods = baseStringCandidates;
```
>candidateMethods [](start = 20, length = 16)

Do we really need to go through the trouble of doing overload resolution again for this case?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                        free();
+                        builderFormatCalls.Free();
+                        builderArguments = default;
+                        return false;
```
> return false; [](start = 23, length = 14)

Dropping useSiteInfo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                    if (insert.Alignment is not null)
+                    {
+                        analyzedArguments.Arguments.Add(insert.Alignment);
+                        analyzedArguments.Names.Add(SyntaxFactory.IdentifierName("alignment"));
```
>SyntaxFactory.IdentifierName("alignment") [](start = 52, length = 41)

I think we should find a way to provide names without synthesizing fake syntax nodes.

333fred replied:
Needs a bit of an involved refactoring, that will have it's own commit.

---
In reply to: [610733589](https://github.com/dotnet/roslyn/pull/52422#discussion_r610733589) [](ancestors = 610733589)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+
+                var argsToParam = overloadResolutionResult.ValidResult.Result.ArgsToParamsOpt;
+
+                _binder.CoerceArguments(overloadResolutionResult.ValidResult, analyzedArguments.Arguments, diagnostics);
```
>CoerceArguments [](start = 24, length = 15)

It feels like we are doing too much manual work (coercing arguments, binding default arguments, etc.) here and creating a maintenance burden for the future in terms of making sure all the checks that we perform for regular call are also performed here and tested here, etc. I think we should use existing methods in the binder to bind the call into a ```BoundCall``` node based on the set of candidates and given arguments. Then we can deconstruct that node into ```MethodArgumentInfo```, but I am not even sure why would we want to do that instead of simply storing the ```BoudCall``` and using it later in the pipeline.

333fred replied:
The main issue here is that we don't have a node to act as a receiver for this call. I suppose I could make a placeholder receiver that we then swap out in lowering for the actual builder instance, but we don't do similar tricks for any of the foreach calls.

---
In reply to: [610745021](https://github.com/dotnet/roslyn/pull/52422#discussion_r610745021) [](ancestors = 610745021)

AlekseyTs replied:
> The main issue here is that we don't have a node to act as a receiver for this call. I suppose I could make a placeholder receiver that we then swap out in lowering for the actual builder instance, but we don't do similar tricks for any of the foreach calls.

For each calls are significantly simpler than this.

---
In reply to: [612035214](https://github.com/dotnet/roslyn/pull/52422#discussion_r612035214) [](ancestors = 612035214,610745021)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                        return false;
+                    }
+
+                    candidateMethods.Add(method);
```
>method [](start = 41, length = 6)

Can we get an extension method here?

333fred replied:
Shouldn't be possible, we have `LookupOptions.MustBeInstance` without `LookupOptions.IncludeExtensionMethods`.

---
In reply to: [610747594](https://github.com/dotnet/roslyn/pull/52422#discussion_r610747594) [](ancestors = 610747594)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:
```diff
@@ -74,6 +73,10 @@ internal ConversionsBase WithNullability(bool includeNullability)
         protected abstract Conversion GetInterpolatedStringConversion(BoundUnconvertedInterpolatedString source, TypeSymbol destination, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo);
 
+#nullable  enable
+        public abstract bool IsApplicableInterpolatedStringBuilderType(BoundUnconvertedInterpolatedString source, TypeSymbol builderType, BindingDiagnosticBag diagnostics, [NotNullWhen(true)] out ImmutableArray<MethodArgumentInfo> builderArguments);
```
>IsApplicableInterpolatedStringBuilderType [](start = 29, length = 41)

It looks like this method is not used to classify conversion, but rather performs some part of binding work and is dependent on a Binder. I think we should move it to the Binder.

333fred replied:
> It looks like this method is not used to classify conversion, but rather performs some part of binding work and is dependent on a Binder. I think we should move it to the Binder.

The reason I have it here because I believe I am going to need it when doing overload resolution, when determining whether an interpolated string expression is convertible to a parameter type.

---
In reply to: [610785056](https://github.com/dotnet/roslyn/pull/52422#discussion_r610785056) [](ancestors = 610785056)

AlekseyTs replied:
>> It looks like this method is not used to classify conversion, but rather performs some part of binding work and is dependent on a Binder. I think we should move it to the Binder.

> The reason I have it here because I believe I am going to need it when doing overload resolution, when determining whether an interpolated string expression is convertible to a parameter type.

I think we should not add the API until then. Also, given what the method is doing, I think it belongs in the Binder.

---
In reply to: [612032667](https://github.com/dotnet/roslyn/pull/52422#discussion_r612032667) [](ancestors = 612032667,610785056)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+            //     We also cannot use this method if the interpolated string appears within a catch filter, as the builder is disposable and we
+            //     cannot put a try/finally inside a filter block.
+            //  3. The string is composed entirely of components that are strings themselves. We can turn this into a single call to string.Concat.
+            //     We prefer the builder over this because the builder can used pooling to avoid new allocations, while this call will potentially
```
>can used [](start = 71, length = 8)

Typo?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                            partsBuilder.AddRange(unconvertedInterpolatedString.Parts, i);
+                        }
+
+                        partsBuilder.Add(insert.Update(newValue, insert.Alignment, insert.Format, insert.Type));
```
>insert.Type [](start = 98, length = 11)

It looks like ```BoundStringInsert.Type``` is always null. Consider adjusting the node definition accordingly so that we wouldn't need to pass the type and there would be no questions if the type should also be adjusted here and in other similar places.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -87,6 +88,234 @@ protected override Conversion GetInterpolatedStringConversion(BoundUnconvertedIn
+                overloadResolutionResult,
+                ref useSiteInfo);
+
+            if (!overloadResolutionResult.Succeeded || overloadResolutionResult.ValidResult.Member.CallsAreOmitted(source.SyntaxTree))
```
>overloadResolutionResult.ValidResult.Member.CallsAreOmitted(source.SyntaxTree) [](start = 55, length = 78)

Why do we even care about this? And, if we do, why not simply report a warning?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                result = constructWithData(
+                    newParts,
+                    new InterpolatedStringBuilderData(
+                        interpolatedStringBuilderType,
```
>interpolatedStringBuilderType [](start = 24, length = 29)

Are we failing to report use-site diagnostics and record dependencies for the type?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                if (!Conversions.IsApplicableInterpolatedStringBuilderType(unconvertedInterpolatedString, interpolatedStringBuilderType, applicableDiagnostics, out var builderArguments))
+                {
+                    applicableDiagnostics.Free();
+                    return false;
```
>return false; [](start = 20, length = 13)

I am not sure whether this is a design question or more an implementation question, but I would prefer the implementation and the rules be simpler. In particular, If the builder type is there, and other conditions not related to the implementation of the builder type are met (lack of awaits, not in catch filter, etc.), I think we should bind interpolation as a builder. Any unexpected situations we run into (lack of overloads or inconsistent overloads, etc.) should be treated as errors and be reported as such. Otherwise, the implementation feels fragile, it is not clear what errors should be treated as real errors and what errors should be ignored along with the type. For consumer, it is hard to figure out why the builder type is getting ignored when it looks like it has everything it needs.

333fred replied:
This will partially depend on the builder definition we decide on in the framework. With the current plan, we should able to support any expression type aside from things that will never be supported no matter whether we do `string.Format` or the builder, like pointer types. If we confirm that design decision tomorrow, I think we can simplify this code.

---
In reply to: [610810182](https://github.com/dotnet/roslyn/pull/52422#discussion_r610810182) [](ancestors = 610810182)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    Debug.Assert(argumentsBuilder.Count == 0);
+
+                    var newPart = currentFormatCall.Arguments[0];
+                    argumentsBuilder.Add(new BoundInterpolatedStringElementPlaceholder(newPart.Syntax, i, newPart.Type));
```
>argumentsBuilder.Add(new BoundInterpolatedStringElementPlaceholder(newPart.Syntax, i, newPart.Type)); [](start = 20, length = 101)

I am not sure why we are doing this. It also looks like an assumption is made that the part and coerced argument are the same thing, whereas I don't think this is necessary the case. I think we either should bind againt a placeholder from the very beginning and then use standard substitution during lowering as we do with other placeholders, or we should leave the argument alone and just lower it normally.

333fred replied:
>  I think we either should bind againt a placeholder from the very beginning

This won't work with target-typed expressions.

---
In reply to: [610901304](https://github.com/dotnet/roslyn/pull/52422#discussion_r610901304) [](ancestors = 610901304)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                            }
+                        }
+
+                        newPart = insert.Update(newPart, insert.Alignment, insert.Format, newPart.Type);
```
>newPart.Type [](start = 90, length = 12)

This looks suspicious because from the current sources it appears that ```BoundStringInsert.Type``` is always supposed to be null.

AlekseyTs replied:
>newPart [](start = 24, length = 7)

It looks like we are confusing aruments with parts.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    }
+                    else
+                    {
+                        Debug.Assert(newPart is BoundLiteral { ConstantValue: { IsString: true } });
```
>Debug.Assert(newPart is BoundLiteral { ConstantValue: { IsString: true } }); [](start = 24, length = 76)

I don't think we can make this assumption.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    partsBuilder.Add(newPart);
+
+                    ReportDiagnosticsIfObsolete(diagnostics, currentFormatCall.Method, currentPart.Syntax, hasBaseReceiver: false);
+                    ReportDiagnosticsIfUnmanagedCallersOnly(diagnostics, currentFormatCall.Method, currentPart.Syntax.Location, isDelegateConversion: false);
```
Another example of a manual work that I think we should avoid doing.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                if (needToCheckConversionToObject)
+                {
+                    objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                    conversionDiagnostics = BindingDiagnosticBag.GetInstance();
```
>GetInstance [](start = 65, length = 11)

Is it necessary to track dependencies?

333fred replied:
I meant to use the input as a template, thanks.

---
In reply to: [610909474](https://github.com/dotnet/roslyn/pull/52422#discussion_r610909474) [](ancestors = 610909474)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                            if (!reported)
+                            {
+                                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(conversionDiagnostics);
+                                var conversion = Conversions.ClassifyConversionFromExpression(value, objectType, ref useSiteInfo);
```
>Conversions.ClassifyConversionFromExpression(value, objectType, ref useSiteInfo); [](start = 49, length = 81)

Is this sufficient to pick up any errors that could come from ```GenerateConversionForAssignment```. Consider using ```GenerateConversionForAssignment``` to match what legacy compiler would do.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    constructorInfo = new MethodArgumentInfo(createMethod, arguments, ArgsToParamsOpt: default, DefaultArguments: BitVector.Empty, Expanded: false);
+
+                    ReportDiagnosticsIfObsolete(diagnostics, createMethod, unconvertedInterpolatedString.Syntax, hasBaseReceiver: false);
+                    ReportDiagnosticsIfUnmanagedCallersOnly(diagnostics, createMethod, unconvertedInterpolatedString.Syntax.Location, isDelegateConversion: false);
```
I don't think we bother performing these checks for well known members.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    diagnostics,
+                    syntax: unconvertedInterpolatedString.Syntax);
+
+                MethodArgumentInfo? constructorInfo = null;
```
>constructorInfo [](start = 36, length = 15)

The name feels confusing, it doesn't store a constructor info.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                        new BoundLiteral(unconvertedInterpolatedString.Syntax, ConstantValue.Create(baseStringLength), intType) { WasCompilerGenerated = true },
+                        new BoundLiteral(unconvertedInterpolatedString.Syntax, ConstantValue.Create(numFormatHoles), intType) { WasCompilerGenerated = true });
+
+                    constructorInfo = new MethodArgumentInfo(createMethod, arguments, ArgsToParamsOpt: default, DefaultArguments: BitVector.Empty, Expanded: false);
```
>constructorInfo = new MethodArgumentInfo(createMethod, arguments, ArgsToParamsOpt: default, DefaultArguments: BitVector.Empty, Expanded: false); [](start = 20, length = 144)

I am not sure why is it necessary to do this work in binding, we could easily synthesize the call in lowering.

333fred replied:
This will need to get more generic to support non-well-known create methods for other builder types. I _could_ special case this instance to not do the work now and do it in lowering instead, but that will just result in more code dealing with the special case down the road and I'd rather avoid it.

---
In reply to: [610915127](https://github.com/dotnet/roslyn/pull/52422#discussion_r610915127) [](ancestors = 610915127)

AlekseyTs replied:
> This will need to get more generic to support non-well-known create methods for other builder types. I could special case this instance to not do the work now and do it in lowering instead, but that will just result in more code dealing with the special case down the road and I'd rather avoid it.

Then I would consider not using WellKnownMembers at all, but rather do the normal binding for a call.

---
In reply to: [612042647](https://github.com/dotnet/roslyn/pull/52422#discussion_r612042647) [](ancestors = 612042647,610915127)

333fred replied:
I don't think this is necessary. This codepath is only ever dealing with one type, and that type needs to have the things we expect or it's malformed. Normal binding does a lot of work we don't care about here.

---
In reply to: [612777789](https://github.com/dotnet/roslyn/pull/52422#discussion_r612777789) [](ancestors = 612777789,612042647,610915127)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+
+                MethodArgumentInfo? disposeInfo = null;
+
+                if (disposeMethod != null)
```
>disposeMethod [](start = 20, length = 13)

Same comments as for constructorInfo

333fred replied:
No longer ncessary.

---
In reply to: [610916036](https://github.com/dotnet/roslyn/pull/52422#discussion_r610916036) [](ancestors = 610916036)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -35,6 +36,108 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+
+            // PROTOTYPE(interp-string): Support optional out param for whether the builder was created successfully
+            // var builder = Construction(baseStringLength, numFormatHoles);
+            var builderConstruction = _factory.Assignment(builderTemp, MakeCallWithNoExplicitArgument(node.InterpolationData.Construction, node.Syntax, expression: null, assertParametersAreOptional: false));
```
>MakeCallWithNoExplicitArgument [](start = 71, length = 30)

Why are we using "WithNoExplicitArgument" helper when we know that there are explicit arguments?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -35,6 +36,108 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                var formatInfo = node.InterpolationData.BuilderFormatCalls[i];
+                var part = node.Parts[i];
+                Debug.Assert(usesBoolReturn == (formatInfo.Method.ReturnType.SpecialType == SpecialType.System_Boolean));
+                Debug.Assert(formatInfo.Arguments[0] is BoundInterpolatedStringElementPlaceholder { Index: var index } && index == i);
```
>BoundInterpolatedStringElementPlaceholder [](start = 56, length = 41)

I am completely missing the benefit of switching this node back and forth

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -35,6 +36,108 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                var visitedPart = VisitExpression(part is BoundStringInsert { Value: var value } ? value : part);
+                var formatCall = MakeCallWithNoExplicitArgument(formatInfo with
+                {
+                    Arguments = formatInfo.Arguments.SetItem(0, visitedPart)
```
>formatInfo.Arguments [](start = 32, length = 20)

Are we failing to lower the arguments except the first one?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -35,6 +36,108 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+
+                // Visit the part, and replace the part placeholder in the TryFormat... call with the replaced part.
+                var visitedPart = VisitExpression(part is BoundStringInsert { Value: var value } ? value : part);
+                var formatCall = MakeCallWithNoExplicitArgument(formatInfo with
```
>MakeCallWithNoExplicitArgument [](start = 33, length = 30)

The use of this method doesn't feel appropriate because it is not meant to be used with arguments. Even if it appears to be working, it is not meant to deal with arguments produced by initial binding.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -35,6 +36,108 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+            // PROTOTYPE(interp-string): This will need to be a lot more complicated to support method parameter scenarios
+            // resultTemp = builder.ToString();
+
+            builderPatternStatements.Add(_factory.Assignment(resultTemp, _factory.InstanceCall(builderTemp, "ToString")));
```
>InstanceCall [](start = 82, length = 12)

It looks like this is the first usage of the helper. Either we should find an alternative, or a very thorough testing should be performed to ensure we get expected behavior in various cases (method is missing, inaccessible, etc.)

AlekseyTs replied:
>_factory.Assignment(resultTemp, _factory.InstanceCall(builderTemp, "ToString")) [](start = 41, length = 79)

Why do we assume the call returns a string?

AlekseyTs commented:
Done with review pass (commit 4), tests are not looked at.

333fred commented:
@AlekseyTs addressed feedback. I'd suggest going commit by commit: commit 5 is the AnalyzedArguments refactoring so we don't have to synthesize syntax, commit 6 moves the method from Conversions to the Binder (with only method accessibility changes), and commit 7 is the rest of the in-place refactoring.

---
In reply to: [817002240](https://github.com/dotnet/roslyn/pull/52422#issuecomment-817002240) [](ancestors = 817002240)

chsienki commented:
@333fred re-looking at this now.

333fred commented:
@AlekseyTs for another review pass.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -476,8 +476,8 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
                     int badArg = firstSupported.Result.BadArgumentsOpt[0];
-                    IdentifierNameSyntax badName = arguments.Names[badArg];
-                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, badName.Location);
+                    Location badName = arguments.Names[badArg].Value.Location;
```
I get the change to not take an actual SyntaxNode, but just using a tuple seems kind of... leaky. I guess it comes down to a team preference of tuples vs dedicated structs.

333fred replied:
A dedicated struct just felt like overkill here.

chsienki commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -773,16 +773,14 @@ private bool HadConstraintFailure(Location location, BindingDiagnosticBag diagno
             // We would not have gotten this error had there not been a named argument.
             Debug.Assert(arguments.Names.Count > badArg);
-            IdentifierNameSyntax badName = arguments.Names[badArg];
+            (string Name, Location Location)? badName = arguments.Names[badArg];
```
Consider just grabbing the .Value upfront, and Asserting .HasValue beforehand.

chsienki commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,17 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>Interpolated string builder method '{0}' has inconsistent return types. Expected to return '{0}'.</value>
+  </data>
+  <data name="ERR_InterpolatedStringBuilderInvalidCreateMethod" xml:space="preserve">
+    <value>Interpolated string builder type '{0}' does not have a valid 'Create' method.</value>
```
We don't have a 'Create' method any more do we?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/AnalyzedArguments.cs:
```diff
@@ -42,18 +39,23 @@ public BoundExpression Argument(int i)
-            IdentifierNameSyntax syntax = Names[i];
-            return syntax == null ? null : syntax.Identifier.ValueText;
+            var nameAndLocation = Names[i];
+            return nameAndLocation is (string name, _) ? name : null;
```
>nameAndLocation is (string name, _) ? name : null [](http://example.com/codeflow?start=19&length=49)

```nameAndLocation?.Name``` feels much simpler and easier to follow.

333fred replied:
It absolutely is, sometimes you get attracted by the new shiny and forget the existing thing that works better.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1612,7 +1612,7 @@
     <Field Name="ConstructorArguments" Type="ImmutableArray&lt;BoundExpression&gt;"/>
-    <Field Name="ConstructorArgumentNamesOpt" Type="ImmutableArray&lt;string&gt;" Null="allow"/>
+    <Field Name="ConstructorArgumentNamesOpt" Type="ImmutableArray&lt;string?&gt;" Null="allow"/>
     <Field Name="ConstructorArgumentsToParamsOpt" Type="ImmutableArray&lt;int&gt;" Null="allow"/>
```
The motivation behind this change isn't obvious. How and why attributes are affected?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1627,7 +1627,7 @@
       <Field Name="Type" Type="TypeSymbol?" Override="true" Null="always"/>
       <Field Name="Arguments" Type="ImmutableArray&lt;BoundExpression&gt;"/>
-      <Field Name="ArgumentNamesOpt" Type="ImmutableArray&lt;IdentifierNameSyntax&gt;" Null="allow"/>
+      <Field Name="ArgumentNamesOpt" Type="ImmutableArray&lt;(string, Location)?&gt;" Null="allow"/>
```
>(string, Location) [](http://example.com/codeflow?start=61&length=18)

Consider adding descriptive names for tuple elements

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution_ArgsToParameters.cs:
```diff
@@ -317,7 +317,7 @@ public ImmutableArray<int> ToImmutableArray()
                     // p is initialized to zero; it is ok for a named argument to "correspond" to
                     // _any_ parameter (not just the parameters past the point of positional arguments)
-                    if (memberParameters[p].Name == name.Identifier.ValueText)
+                    if (memberParameters[p].Name == name.Value.Name)
```
>Value [](http://example.com/codeflow?start=57&length=5)

```GetValueOrDeault()``` is more efficient because it avoids an unnecessary null check.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -476,8 +476,8 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
                     int badArg = firstSupported.Result.BadArgumentsOpt[0];
-                    IdentifierNameSyntax badName = arguments.Names[badArg];
-                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, badName.Location);
+                    Location badName = arguments.Names[badArg].Value.Location;
```
>Value [](http://example.com/codeflow?start=63&length=5)

GetValueOrDefault?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -773,16 +773,14 @@ private bool HadConstraintFailure(Location location, BindingDiagnosticBag diagno
                 ErrorCode.ERR_NamedArgumentUsedInPositional,
-                new object[] { badName.Identifier.ValueText },
-                symbols), location);
+                new object[] { badName.Value.Name },
```
>Value [](http://example.com/codeflow?start=39&length=5)

GetValueOrDefault?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -773,16 +773,14 @@ private bool HadConstraintFailure(Location location, BindingDiagnosticBag diagno
-                new object[] { badName.Identifier.ValueText },
-                symbols), location);
+                new object[] { badName.Value.Name },
+                symbols), badName.Value.Location);
```
>Value [](http://example.com/codeflow?start=34&length=5)

GetValueOrDefault?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -8018,7 +8011,7 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
                         ? ErrorCode.ERR_ImplicitRangeIndexerWithName
                         : ErrorCode.ERR_ImplicitIndexIndexerWithName,
-                    arguments.Names[0].GetLocation());
+                    arguments.Names[0].Value.Location);
```
>Value [](http://example.com/codeflow?start=39&length=5)

GetValueOrDefault?

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,17 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <comment>void and bool are keywords</comment>
+  </data>
+  <data name="ERR_InterpolatedStringBuilderMethodReturnInconsistent" xml:space="preserve">
+    <value>Interpolated string builder method '{0}' has inconsistent return types. Expected to return '{0}'.</value>
```
>types [](http://example.com/codeflow?start=76&length=5)

Typo "types" vs. "type"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+
+                    // If we were able to actually bind the Append call, then just update the current part to be the call.
+                    // Otherwise, there was some kind of error, and we just want to keep the original part. We'll update
+                    // the HasErrors to be true, but we don't want IOperation to have to try and find the original node
```
>the HasErrors to be true [](http://example.com/codeflow?start=23&length=24)

Are we taking advantage of this in some way? Otherwise, it feels like the error flag should be on the interpolated string result node that we create rather than on a particular part.

AlekseyTs replied:
It looks like Alignment can be a complex expression consisting of multiple nodes. It looks like all of them will be duplicated. Let's discuss this issue offline.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                                if (conversionDiagnostics.HasAnyErrors())
+                                {
+                                    CheckFeatureAvailability(value.Syntax, MessageID.IDS_FeatureImprovedInterpolatedStrings, diagnostics);
+                                    conversionDiagnostics.Clear();
```
>conversionDiagnostics.Clear(); [](http://example.com/codeflow?start=36&length=30)

Perhaps should clear unconditionally

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                                if (conversionDiagnostics.HasAnyErrors())
+                                {
+                                    CheckFeatureAvailability(value.Syntax, MessageID.IDS_FeatureImprovedInterpolatedStrings, diagnostics);
+                                    conversionDiagnostics.Clear();
```
>conversionDiagnostics.Clear(); [](http://example.com/codeflow?start=36&length=30)

Perhaps should clear unconditionally

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                if (needToCheckConversionToObject)
+                {
+                    objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                    conversionDiagnostics = BindingDiagnosticBag.GetInstance(template: diagnostics);
```
>template: diagnostics [](http://example.com/codeflow?start=77&length=21)

It looks like we don't need to track dependencies after all, regardless of what the ```diagnosics``` is doing.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                return (ImmutableArray<BoundExpression>.Empty, false);
+            }
+
+            var implicitBuilderReceiver = new BoundInterpolatedStringBuilderPlaceholder(source.Syntax, builderType);
```
>BoundInterpolatedStringBuilderPlaceholder [](http://example.com/codeflow?start=46&length=41)

Make compiler generated?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+            var expression = MakeInvocationExpression(syntax, typeExpression, "Create", args, diagnostics);
+            if (expression is BoundCall call && call.Type.Equals(builderType, TypeCompareKind.AllIgnoreOptions))
+            {
+                return call;
```
>call [](http://example.com/codeflow?start=23&length=4)

Is there a guarantee that the node is marked as compiler generated? Otherwise we should consider marking it here.

333fred replied:
There is. `MakeInvocationExpression` is specifically for compiler-synthesized calls and unconditionally marks the invocation it returns as compiler-generated.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+            {
+                Debug.Assert(part is BoundLiteral or BoundStringInsert);
+                argumentsBuilder.Clear();
+                parameterNamesAndLocationsBuilder.Clear();
```
>parameterNamesAndLocationsBuilder.Clear(); [](http://example.com/codeflow?start=16&length=42)

It looks like builders are cleared at the end of each iteration

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                }
+
+                var call = MakeInvocationExpression(part.Syntax, implicitBuilderReceiver, methodName, arguments, diagnostics, names: parameterNamesAndLocations, searchExtensionMethodsIfNecessary: false);
+                builderAppendCalls.Add(call);
```
>call [](http://example.com/codeflow?start=39&length=4)

Is there a guarantee that the node is marked as compiler generated? Otherwise we should consider marking it here.

333fred replied:
There is. `MakeInvocationExpression` is specifically for compiler-synthesized calls and unconditionally marks the invocation it returns as compiler-generated.

AlekseyTs replied:
> There is. MakeInvocationExpression is specifically for compiler-synthesized calls and unconditionally marks the invocation it returns as compiler-generated.

It looks like there is a an early return from that method that doesn't do that.

333fred replied:
That does feel more like a bug in MakeInvocationExpression, but alright, I can adjust.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    else if (builderPatternExpectsBool != methodReturnsBool)
+                    {
+                        // Interpolated string builder method '{0}' has inconsistent return types. Expected to return '{0}'.
+                        var expected = builderPatternExpectsBool == true ? Compilation.GetSpecialType(SpecialType.System_Boolean) : Compilation.GetSpecialType(SpecialType.System_Object);
```
>System_Object [](http://example.com/codeflow?start=171&length=13)

Is this accurate?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -108,7 +173,24 @@ public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
+                    builderPatternExpressions.Add(_factory.AssignmentExpression(resultTemp, BoundCall.Synthesized(node.Syntax, builderTemp, toStringAndClear)));
+                }
+
+                return _factory.Sequence(ImmutableArray.Create(resultSymbol, builderTemp.LocalSymbol), builderPatternExpressions.ToImmutableAndFree(), resultTemp);
```
>resultTemp [](http://example.com/codeflow?start=151&length=10)

What is the motivation to use this temp? It looks like we could simply use the call at this position.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    {
+                        Debug.Assert(call.Arguments.Length > 0);
+                        newPart = currentPart is BoundStringInsert si
+                            ? si.Update(call, si.Alignment, si.Format, isInterpolatedStringBuilderAppendCall: true)
```
>si.Update(call, si.Alignment, si.Format, isInterpolatedStringBuilderAppendCall: true) [](http://example.com/codeflow?start=30&length=85)

Is there a specific reason why we need to preserve BoundStringInsert node in the tree? It looks like we are creating a node that refers to Alignment and Format nodes twice and none of them marked as compiler generated. This is likely to cause confusion for SemanticModel. Are we testing getting IOperation and type info for the nodes?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -35,6 +37,69 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+                var appendCall = (BoundCall)(currentPart is BoundStringInsert si ? si.Value : currentPart);
+                Debug.Assert(usesBoolReturn == (appendCall.Method.ReturnType.SpecialType == SpecialType.System_Boolean));
+
+                var rewrittenArgs = VisitList(appendCall.Arguments);
```
>VisitList [](http://example.com/codeflow?start=36&length=9)

Are we doing this manually only because we need to substitute receiver? Consider updating the original node with new receiver (asserting that we are substituting expected node) and then calling VisitCall

333fred replied:
Partially that, but also partially because the real receiver was created in lowering and `VisitCall` will visit it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1973,34 +1973,70 @@ internal IOperation CreateBoundTupleOperation(BoundTupleExpression boundTupleExp
+            var value = (boundStringInsert.IsInterpolatedStringBuilderAppendCall, boundStringInsert.Value) switch
+            {
+                (false, var v) => v,
+                (true, BoundCall call) => call.Arguments[0],
```
>call.Arguments[0] [](http://example.com/codeflow?start=42&length=17)

In general this can be a conversion on top of the actual value from the hole. I think we need a more robust way to locate the original node. Alternatively, we should just expose the calls in a new kind of node.

333fred replied:
I'm going to revert the IOp changes for now and just disable the new tests in the test hook run.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1973,34 +1973,70 @@ internal IOperation CreateBoundTupleOperation(BoundTupleExpression boundTupleExp
+                            Debug.Assert(usesBuilderPattern);
+                            Debug.Assert(call.Arguments.Length == 1);
+
+                            var currentArg = call.Arguments[0];
```
>call.Arguments[0] [](http://example.com/codeflow?start=45&length=17)

This could be a conversion on top of the original node.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -1195,5 +1197,1805 @@ static void Main()
+
+            var comp1 = CreateCompilation(interpolatedStringBuilder);
+
+            foreach (var reference in new[] { comp1.EmitToImageReference(), comp1.EmitToPortableExecutableReference() })
```
>EmitToPortableExecutableReference [](http://example.com/codeflow?start=82&length=33)

It doesn't look like there is a difference by comparison to the first item. Did you mean to use ToMetadataReference?

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,17 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+  <data name="IDS_FeatureImprovedInterpolatedStrings" xml:space="preserve">
+    <value>interpolated string builders</value>
+  </data>
+  <data name="ERR_InterpolatedStringBuilderMethodReturnMalformed" xml:space="preserve">
```
>ERR_InterpolatedStringBuilderMethodReturnMalformed [](http://example.com/codeflow?start=14&length=50)

It looks like there is no test coverage for this error

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,17 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>Interpolated string builder method '{0}' is malformed. It does not return 'void' or 'bool'.</value>
+    <comment>void and bool are keywords</comment>
+  </data>
+  <data name="ERR_InterpolatedStringBuilderMethodReturnInconsistent" xml:space="preserve">
```
>ERR_InterpolatedStringBuilderMethodReturnInconsistent [](http://example.com/codeflow?start=14&length=53)

It looks like there is no test coverage for this error

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -1195,5 +1197,1805 @@ static void Main()
+                // Console.WriteLine($"{s}");
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "{s}").WithArguments("S").WithLocation(5, 21)
+            );
+        }
```
>} [](http://example.com/codeflow?start=8&length=1)

Are we testing:
- scenario when AppendLiteral uses user-defined conversion
- SemanticModel APIs
- IOperation tree

333fred replied:
I'll add the first one. The latter 2 will come at a later time.

AlekseyTs commented:
Done with review pass (commit 7)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                    else
+                    {
+                        Debug.Assert(currentFormatCall.HasErrors);
+                        newPart = currentPart.WithHasErrors();
```
>currentPart [](http://example.com/codeflow?start=34&length=11)

If we are dropping the result of binding a call, It feels like at least we need to snap this node to natural type for error recovery.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [828889879](https://github.com/dotnet/roslyn/pull/52422#issuecomment-828889879) [](http://example.com/codeflow?ancestors=828889879,828889879)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -1195,5 +1197,2044 @@ static void Main()
+            var source = @"
+using System;
+
+Console.WriteLine($""Text{1}"");
```
>$""Text{1}"" [](http://example.com/codeflow?start=18&length=12)

Is it possible to test scenario when AppendLiteral is called second? For example, ```$""{1}Text""```?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -1195,5 +1197,2044 @@ static void Main()
+            var source = @"
+using System;
+
+Console.WriteLine($""Text{1}"");
```
>$""Text{1}"" [](http://example.com/codeflow?start=18&length=12)

Same question.

AlekseyTs reviewed (approved):
LGTM (commit 8) with a follow-up.

chsienki commented:
LGTM

AlekseyTs reviewed (approved):
LGTM (commit 11)
